### ISSUE TITLE
bump FutureSoots Java representation to Modules

### ISSUE THREAD

**Issue by swissiety** [soot-oss/SootUp#185]
Resources:
https://openjdk.java.net/projects/jigsaw/quick-start
https://openjdk.java.net/projects/jigsaw/spec/sotms/
https://www.baeldung.com/java-9-modularity
https://www.informatik-aktuell.de/entwicklung/programmiersprachen/java-9-das-neue-modulsystem-jigsaw-tutorial.html#c20123

## TODO
depends on #186
- [x] *replace* <java9 Signatures with a module (prefix) - or *add* another `Signature` SubClass
> overhead for <java9 analyses but no additional/seperate handling for backwards compatible designed stuff, another `(JavaModule)View` would make it possible though, but when introduce the change it if not now, even it shouldn't be a problem to keep the current non-module api while extending it(?)

> selection which Java*View is used can be determined by the Builder in JavaProject.. compose JavaModuleView with JavaView when necessary.
- [x]  change current resolving mechanism of w/AsmModuleClassSource

- [x] incorporate a catchall default module -> the unnamed module
- [x] adapt getClass() : needs to know which module refers to the class
- [x] adapt getClasses() like above
- [x] resolve dependent class in the correct submodule (make it lazy / cache already calculated deps)

#### information from module-info.class
- [x] a module-info without directives is valid
- [x] requires 
- [x] requires transitive
- [x] exports
- [x] exports ... to ...
- [x] uses 
- [x] provides ... with ...

allow reflection (*nicetohave*):
  - [x] open module 
  - [x] opens  directive
  - [x] opens ... to ... directive

### Tests
resources: https://github.com/accso/java9-jigsaw-examples/tree/master/jigsaw-examples
- [x] exploded vs jar
> see structure in ide, no zip layer ->faster vs no easy messing around with modules class files, less "noise"
- [x] mandatory dependency on java.base 
- [x] explicit modules (no access no unnamed module  )
  - [x] system modules i.e. java.base
  - [ ] application modules (no need to specify the entrypoint for execution)
  > think for convenience about Entrypoints in `View`
- [x] automatic module ( no module-info; "normal" jar; is an open module; access to every other module on module path; default requires on unnamed module; exports: *)
	- [x] build name from filename
	- [x] get name from manifest
- [x] unnamed module (the catchall module for everything from **class**path: exports: *; requires:* )
- [ ] test failing on split packages (package(signature)s are not disunct across multiple modules)
- [ ] split packages https://www.logicbig.com/tutorials/core-java-tutorial/modules/split-packages.html

### think about
- [ ] serialisierung/parsing via jimple
	- [x] SIgnatures
	- [ ] JavaModuleInfo
